### ISSUE TITLE
Rule suggestion: restrict + operands

### ISSUE THREAD

**Issue by myitcv** [palantir/tslint#697]
With reference to the [binary `+` operator in the spec](https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#4192-the--operator)

I find it rather disconcerting that this is allowed:

``` typescript
let x = 5 + "test";
let y = true + "test";
```

Would there be interest in a rule that forces the two operands to `+` to either:
- both be of type `string`
- both be of type `number`

and fail otherwise?

Do we have the necessary type information available during linting to do this?

**Comment by myitcv** [comment]
I've made a start, but just [run into a problem](https://github.com/Microsoft/TypeScript/issues/5036)

**Comment by adidahiya** [comment]
sounds like a good lint rule, thanks for looking into this!

**Comment by adidahiya** [comment]
TSLint currently only gets type information on a per-file basis, so to do this properly, I think we're blocked on #680

**Comment by myitcv** [comment]
Ah. Is that why the `TypeChecker` returns `Any`?

**Comment by myitcv** [comment]
Sorry, to be more explicit: that question relates to the [TypeScript issue I created](https://github.com/Microsoft/TypeScript/issues/5036)

**Comment by adidahiya** [comment]
No, I'm not sure why the type checker returns `any` in that issue you created -- intuitively it feels like the checker has enough information about the built-in `.toString()` to report the correct return type.

Perhaps we're not instantiating the checker correctly?

**Comment by myitcv** [comment]
Hmm, having done a quick test I think it is related. In the following example the `TypeChecker` (in the way we are using it with a single file) reports the type of `Banana()` is `Any`:

```typescript
import {Banana} from "./somewhere_else";

// Banana has type signature: Banana(): string

let y = "test" + Banana();
```

My guess would be that when it hits an identifier defined outside of the scope of what is being checked it assumes an `Any` type

**Comment by JKillian** [comment]
I'm not sure if this is accurate, but possibly since `lib.core.d.ts` isn't included in the program, it has no awareness of core typings.

**Comment by myitcv** [comment]
@JKillian exactly. So #680 will definitely allow us to write such rules. Look forward to it landing!

**Comment by alexeagle** [comment]
This one is approved in `tsc` now. https://github.com/Microsoft/TypeScript/issues/7989#issuecomment-224775260

**Comment by myitcv** [comment]
@alexeagle but only if you use `--strictNullChecks`

**Comment by adidahiya** [comment]
@alexeagle thanks for the link. Agreed with @myitcv that it can still be a TSLint rule that is a more granular opt-in than `--strictNullChecks` in the compiler.

**Comment by OliverJAsh** [comment]
Can this be closed now we have https://palantir.github.io/tslint/rules/restrict-plus-operands/?
